website/static/js/koHelpers.js

```diff
@@ -205,6 +207,141 @@ ko.bindingHandlers.fadeVisible = {
     }
 };
 
+// Run Bootstrap tooltip JS automagically
+// http://getbootstrap.com/javascript/#tooltips
+ko.bindingHandlers.tooltip = {
+    init: function(el, valueAccessor) {
+        var params = valueAccessor();
+        $(el).tooltip(params);
+    }
+};
+// Attach view model logic to global keypress events
+ko.bindingHandlers.onKeyPress = {
+    init: function(el, valueAccessor) {
+        $(window).keydown(function(e) {
+            var params = valueAccessor();
+            if (e.keyCode === params.keyCode) {
+                params.listener(e);
+            }
+        });
+    }
+};
+
+/* A binding handler to convert lists into formatted lists, e.g.:
+ * [dog] -> dog
+ * [dog, cat] -> dog and cat
+ * [dog, cat, fish] -> dog, cat, and fish
+ *
+ * This handler should not be used for user inputs.
+ *
+ * Example use:
+ * <span data-bind="listing: {data: ['Alpha', 'Beta', 'Gamma'],
+ *                            map: function(item) {return item.charAt(0) + '.';}}"></span>
+ * yields
+ * <span ...>A., B., and G.</span>
+ */
+ko.bindingHandlers.listing = {
+    update: function(element, valueAccessor, allBindings, viewModel, bindingContext) {
+        var value = valueAccessor();
+        var valueUnwrapped = ko.unwrap(value);
+        var map = valueUnwrapped.map || function(item) {return item;};
+        var data = valueUnwrapped.data || [];
+        var keys = [];
+        if (!Array.isArray(data)) {
+            keys = Object.keys(data);
+        }
+        else {
+            keys = data;
+        }
+        var index = 1;
+        var list = ko.utils.arrayMap(keys, function(key) {
+            var ret;
+            if (index === 1){
+                ret = '';
+            }
+            else if (index === 2){
+                if (valueUnwrapped.length === 2) {
+                    ret = ' and ';
+                }
+                else {
+                    ret = ', ';
+                }
+            }
+            else {
+                ret = ', and ';
+            }
+            ret += map(key, data[key]);
+            index++;
+            return ret;
+        }).join('');
+        $(element).html(list);
```

Comment:
Why does this set `.html()` instead of `.text()`?

If the strings in `data` contain malicious user entered HTML, it seems that this binding would allow insertion of that text onto the page directly as code. We try to avoid this if at all possible; some HTML escaping functions are available if needed.